Flaky test `TestAuthLeaseTimeToLive`

### Which Github Action / Prow Jobs are flaking?

e2e-386

### Which tests are flaking?

`TestAuthLeaseTimeToLive`

### Github Action / Prow Job link

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/etcd-io_etcd/18574/pull-etcd-e2e-386/1834665788858961920

### Reason for failure (if possible)

```
auth_test.go:847: 
        	Error Trace:	/home/prow/go/src/github.com/etcd-io/etcd/tests/common/auth_test.go:847
        	            				/home/prow/go/src/github.com/etcd-io/etcd/tests/framework/testutils/execute.go:38
        	            				/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.7.linux-amd64/src/runtime/asm_386.s:1363
        	Error:      	An error is expected but got nil.
        	Test:       	TestAuthLeaseTimeToLive
        	Messages:   	test-user must not be able to access to the lease, because it's attached to the key bar
```

### Anything else we need to know?

_No response_

While preparing to run the tests with `stress` I ran them a few times manually (in e2e mode) and I guess I was lucky and saw this failure (same as described in this issue):
```
➜  common git:(issue-18585) go test -v -count 1 -run "^TestAuthLeaseTimeToLive$" -tags=e2e 
=== RUN   TestAuthLeaseTimeToLive
    before.go:36: Changing working directory to: /var/folders/p4/ddf4p2tn7n5_62ynylp1hbf80000gn/T/TestAuthLeaseTimeToLive870828400/001
    logger.go:146: 2024-09-30T10:02:31.369+0200 INFO    starting server...      {"name": "TestAuthLeaseTimeToLive-test-0"}
    logger.go:146: 2024-09-30T10:02:31.369+0200 INFO    spawning process        {"args": ["/Users/thomas/Documents/github.com/ghouscht/etcd/bin/etcd", "--name=TestAuthLeaseTimeToLive-test-0", "--listen-client-urls=http://localhost:20000", "--advertise-client-urls=http://localhost:20000", "--listen-peer-urls=http://localhost:20001", "--initial-advertise-peer-urls=http://localhost:20001", "--initial-cluster-token=new", "--data-dir", "/var/folders/p4/ddf4p2tn7n5_62ynylp1hbf80000gn/T/TestAuthLeaseTimeToLive870828400/002", "--snapshot-count=10000", "--strict-reconfig-check=false", "--auth-token=jwt,pub-key=/Users/thomas/Documents/github.com/ghouscht/etcd/tests/fixtures/server.crt,priv-key=/Users/thomas/Documents/github.com/ghouscht/etcd/tests/fixtures/server.key.insecure,sign-method=RS256,ttl=1s", "--initial-cluster-token=new", "--initial-cluster=TestAuthLeaseTimeToLive-test-0=http://localhost:20001", "--initial-cluster-state=new"], "working-dir": "/private/var/folders/p4/ddf4p2tn7n5_62ynylp1hbf80000gn/T/TestAuthLeaseTimeToLive870828400/001", "name": "TestAuthLeaseTimeToLive-test-0", "environment-variables": ["PATH=/opt/homebrew/Cellar/go/1.23.1/libexec/bin:/Users/thomas/go/bin:/Users/thomas/.krew/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/thomas/go/bin:/Users/thomas/.krew/bin:/Applications/iTerm.app/Contents/Resources/utilities", "ETCD_UNSUPPORTED_ARCH=arm64", "ETCD_VERIFY=all"]}
    logger.go:146: 2024-09-30T10:02:31.978+0200 INFO    started server. {"name": "TestAuthLeaseTimeToLive-test-0", "pid": 72953}
    auth_test.go:847: 
                Error Trace:    /Users/thomas/Documents/github.com/ghouscht/etcd/tests/common/auth_test.go:847
                                                        /Users/thomas/Documents/github.com/ghouscht/etcd/tests/framework/testutils/execute.go:38
                                                        /opt/homebrew/Cellar/go/1.23.1/libexec/src/runtime/asm_arm64.s:1223
                Error:          An error is expected but got nil.
                Test:           TestAuthLeaseTimeToLive
                Messages:       test-user must not be able to access to the lease, because it's attached to the key bar
    logger.go:146: 2024-09-30T10:02:33.098+0200 INFO    closing test cluster...
    logger.go:146: 2024-09-30T10:02:33.098+0200 INFO    stopping server...      {"name": "TestAuthLeaseTimeToLive-test-0"}
    logger.go:146: 2024-09-30T10:02:33.134+0200 INFO    stopped server. {"name": "TestAuthLeaseTimeToLive-test-0"}
    logger.go:146: 2024-09-30T10:02:33.134+0200 INFO    closing server...       {"name": "TestAuthLeaseTimeToLive-test-0"}
    logger.go:146: 2024-09-30T10:02:33.134+0200 INFO    removing directory      {"data-dir": "/var/folders/p4/ddf4p2tn7n5_62ynylp1hbf80000gn/T/TestAuthLeaseTimeToLive870828400/002"}
    logger.go:146: 2024-09-30T10:02:33.135+0200 INFO    closed test cluster.
--- FAIL: TestAuthLeaseTimeToLive (1.77s)
FAIL
exit status 1
FAIL    go.etcd.io/etcd/tests/v3/common 2.133s
```

However, running the tests in `e2e` mode with `stress` is not possible without modification because each test run tries to start its own etcd server, binding the same port which obviusly doesn't work. In `integration` mode I wasn't able to reproduce this issue (8 parallel runs, >10k executed with 0 failures).

To be able to run the tests in e2e mode with stress I made the following modifications in `tests/framework/e2e/e2e.go` to "randomize" the ports of the etcd server (obviously not a very good solution but good enough as a quick fix) :

```patch
 func (e e2eRunner) NewCluster(ctx context.Context, t testing.TB, opts ...config.ClusterOption) intf.Cluster {
        cfg := config.NewClusterConfig(opts...)
+
+       lis, err := net.Listen("tcp", ":0")
+       if err != nil {
+               panic(err)
+       }
+       lis.Close()
+
+       _, port, err := net.SplitHostPort(lis.Addr().String())
+       if err != nil {
+               panic(err)
+       }
+
+       t.Logf("using base port: %s", port)
+
+       p, _ := strconv.Atoi(port)
+
        e2eConfig := NewConfig(
                WithClusterSize(cfg.ClusterSize),
                WithQuotaBackendBytes(cfg.QuotaBackendBytes),
                WithStrictReconfigCheck(cfg.StrictReconfigCheck),
                WithAuthTokenOpts(cfg.AuthToken),
                WithSnapshotCount(cfg.SnapshotCount),
+               WithBasePort(p),
        )
 
        if cfg.ClusterContext != nil {
```

In 1021 runs, I saw 37 failures in total, 3 of them were due to the described issue and the remaining 34 were due to port conflicts (race condition beteween asking the kernel for a free port and actually using it in the parallel test runs). **So that is roughly a 0.3% failure rate for the given test**

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -62,14 +62,16 @@ For more, because etcd uses Kubernetes' prow infrastructure to run CI jobs, the 
 If you find any flaky tests on testgrid, please
 
 1. Check [existing issues](https://github.com/etcd-io/etcd/issues?q=is%3Aissue+is%3Aopen+label%3Atype%2Fflake) to see if an issue has already been opened for this test. If not, open an issue with the `type/flake` label.
-2. Try to reproduce the flaky test on your machine via `stress`, for example, to reproduce the failure of `TestPeriodicSkipRevNotChange`:
+2. Try to reproduce the flaky test on your machine via [`stress`](https://pkg.go.dev/golang.org/x/tools/cmd/stress), for example, to reproduce the failure of `TestPeriodicSkipRevNotChange`:
 
 ```bash
+# install the stress utility
+go install golang.org/x/tools/cmd/stress@latest
 cd server/etcdserver/api/v3compactor
 # compile the test
-go test -v -c -count 1 -run "^TestPeriodicSkipRevNotChange$"
+go test -v -c -count 1
 # run the compiled test file using stress
-stress -p=8 ./v3compactor.test
+stress -p=8 ./v3compactor.test -test.run “^TestPeriodicSkipRevNotChange$”
 ```
 3. Fix it.
 

